src/FluentValidation/DefaultValidatorOptions.cs

```diff
@@ -352,6 +352,41 @@ public static class DefaultValidatorOptions {
 			return rule.Configure(config => config.CurrentValidator.GetMetadata().Severity = severity);
 		}
 
+		public static IRuleBuilderOptions<T, TProperty> OnFailure<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule, Action<T> onFailure) {
+			return rule.Configure(config => config.ReplaceValidator(config.CurrentValidator, new OnFailureValidator<T, TProperty>(config.CurrentValidator, onFailure)));
+		}
+
+		/// <summary>
+		/// Specifies custom method that will be called when specific rule fails
+		/// </summary>
+		/// <typeparam name="T"></typeparam>
+		/// <typeparam name="TProperty"></typeparam>
+		/// <param name="rule"></param>
+		/// <param name="onFailure"></param>
+		/// <returns></returns>
+		public static IRuleBuilderOptions<T, TProperty> OnFailure<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule, Action<T, PropertyValidatorContext> onFailure) {
+			return rule.Configure(config => config.ReplaceValidator(config.CurrentValidator, new OnFailureValidator<T, TProperty>(config.CurrentValidator, onFailure)));
+		}
+
+		public static IRuleBuilderOptions<T, TProperty> OnFailure<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule, Action<T, PropertyValidatorContext, string> onFailure, string errorMessage) {
+			return rule.OnFailure(onFailure, errorMessage, null as object[]);
+		}
+
+		public static IRuleBuilderOptions<T, TProperty> OnFailure<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule, Action<T, PropertyValidatorContext, string> onFailure, string errorMessage, params object[] messageArgs) {

```

Comment:
Remove this overload. Messages with custom args provided by `params object[] messageArgs)`  has been deprecated for quite some time. Instead, we would usually provide a callback that allows a user to make a call to `string.Format` if they want to. Eg:

```
OnFailure<T,TProperty>(this IRuleBuilderOptions<T, TProperty> rule, Action<T, PropertyValidatorContext, string> onFailure, Func<T, string> errorMessageBuilder) 
```

If the user wants to supply custom args for building the message, then they'd then specifify something like this for errorMessageBuilder:
```
x => string.Format("My Message", arg1, arg2);
```

This is how `WithMessage`, `WithLocalizedMessage` etc have worked since 7.x. 

The OnFailure validator would take a `Func<string>` parameter in its ctor instead of just a `string`.

Reply:
This isn't used anymore - I've simplified usage.